### ISSUE TITLE
Incorrect logged DbCommand executed time

### ISSUE THREAD

**Issue by peter-bozovic** [dotnet/efcore#20424]
I have discovered while trying to debug performance issues that the logged time of query execution is invalid.
The console says "Excecuted DbCommand (2ms) while i know the execution of my query took more than 10 seconds, reported by the SQL Server Profiler.
This data is also sent to Application Insights and since it's incorrect it's now harder to identify slow requests and try to fix them (This was not the issue in the 2.2 version)

### Steps to reproduce
Have some slow query executed and compare execution time logged by the application with the execution time provided by the SQL Server Profiler

### Further technical details

EF Core version: 3.1.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1.2
Operating system: Windows 10
IDE: Visual Studio 2019 16.5.0

**Comment by ajcvickers** [comment]
@peter-bozovic It's important to understand that the time reported by EF is the time it takes to send the query to the server and get back the data reader. It does not include the time for the results to be streamed and consumed by the application, because reading is deferred until the application initiates it.

I'm not sure what time is reported by the SQL Server Profiler, but it may be different from the scope of what EF is reporting. Alternatively, this could be a bug, but it would be good to make sure we're comparing apples to apples before we start digging.

**Comment by peter-bozovic** [comment]
Hello @ajcvickers, thanks for your quick reply

You might be right since the query is returning alot of data which takes time to be downloaded.

I have noticed the slow loading times after the upgrade to EF 3.1 which now transforms .Includes into single big query. Instead of few queries that were returning hudreds of rows, now it's building huge queries that are returning hundreds of thousands of rows with all the columns from all the tables (and I guess all that now takes too much time to download, but also to process when building objects.)

Since my logs don't report exact execution and download times it was very hard to identify which ones were causing problems ...

In the meantime I have tweaked logs to get more details of what is going on:
- First there is CommandExecuting
- Few miliseconds later CommandExecuted (which has duration time of few miliseconds)
- And ten seconds later DataReaderDisposing (that doesn't have duration time)

So I guess that logs with times between CommandExecuted and DataReaderDisposing would have been useful. Maybe for some future relase have another telemetry that can send also download duration times so that those can be identified as potential issue and tracked in logs and AI ?

I'm in the process of refactoring all this to remove Includes and use separate queries, but strange how something that was feature in EF Core v1 became wrong practice few versions later and we now have to rework our business logic that worked just fine.

**Comment by smitpatel** [comment]
> can send also download duration times

Also consider that user can start the query and stop enumerating it for long time and then resumes. Any other processing happening during the enumeration would also creep in the time calculation.

**Comment by peter-bozovic** [comment]
So by "user" you mean "EF Core" ?
Because when I do .ToList() .AsEnumerable() or any other command that executes the query it's the EF Core that downloads the data and builds the objects before they are returned to the user.
So this can be considered as something like "Rendering time" which combines the data size and object creation time ? Or is data downloaded first, and then rendered into objects ?

Another thing noticed when inspecing the SQL Server Profiler:
Some queries are executed as `exec sp_executesql N'SELECT ...'` and those report only "RPC:Completed" with duration time.
While this query where I don't see long duration time is executed as "SQL:BatchStarting" and then "SQL;BatchCompleted" with 10 seconds duration time.
I don't know how EF decides which one to use and why , but maybe if all queries were executed as 'exec sp_executesql' the duration time would be the correct one. And I'm sure there is good reason to use SQL Batches but in that case it would be good to have the duration time for those too.

**Comment by smitpatel** [comment]
> So by "user" you mean "EF Core" ?

I meant the user or the application. When you do `ToList` since the list needs to contain all the results, EF Core will build all the object to put in to list. But when you do `AsEnumerable` you are getting an enumerable which is deferred execution. Due to deferred execution which is also supported by DbDataReader, EF Core reads only necessary rows and build objects to give you next item in the enumerable. Hence there is a break between returning each result to user. 
e.g.
```C#
foreach (var item in query)
{
    DoSomething(item); //takes 10 secs to finish
}
```
The code flow would be, EF Core building 1 `item` and then `DoSomething` is called. Once `DoSomething` finishes and foreach reaches next iteration of loop, EF Core will build another item.
In expression tree, EF Core has no way of knowing if you are enumerating it or putting it in ToList.

**Comment by peter-bozovic** [comment]
Nice to know !
This indeed can be useful when you need to process alot of records and you want them as soon as they are available.
And in that case I agree you wound't be able to distinguish if the time spent was within db/ef or in the processing the user was doing on each record.

Still, having this time between the command executing and closing would have been useful for performances troubleshooting.

(In my use case i'm using ToList() and I think I never tried AsEnumerable(), didn't know it was handled differently)

**Comment by ajcvickers** [comment]
The team discussed this and agreed that the value here is low since the application can easily time full execution and consumption of the query, and the potential for any number to be confusing is quite high since any amount of work could be going on in the application. Therefore, on balance, we're not going to measure the time to before the reader is disposed.
